### PR TITLE
[tune] Update trainable docs and support hparams

### PR DESCRIPTION
**Pull request by richardliaw**
## Why are these changes needed?
Trainable _save and _restore errors are too common.

Also, it should be very easy to support hparams tb logging.

Also, add: https://forms.gle/7u5eH1avbTfpZ3dE6

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/tune/trainable.py [hunk 1]
```diff
@@ -462,20 +459,22 @@ def _train(self):
 
         raise NotImplementedError
 
-    def _save(self, checkpoint_dir):
+    def _save(self, tmp_checkpoint_dir):
         """Subclasses should override this to implement save().
 
+        Do not rely on absolute paths in the implementation of `_save` and
+        `_restore`. Use `validate_save_restore(use_object_store=True)`
+        to catch _save/_restore errors before execution.
```
Comment (richardliaw) [inline python/ray/tune/trainable.py]: cc @ericl, an attempt to avoid user errors...

File: doc/source/tune-usage.rst [hunk 2]
```diff
@@ -4,26 +4,30 @@ Tune User Guide
...
+Tune takes a user-defined Python function or class and evaluates it on a set of hyperparameter configurations.
+
+Each hyperparameter configuration evaluation is called a *trial*, and multiple trials are run in parallel. Configurations are either generated by Tune or drawn from a user-specified **search algorithm**. The trials are scheduled and managed by a **trial scheduler**.
 
-Tune schedules a number of *trials* in a cluster. Each trial runs a user-defined Python function or class and is parameterized either by a *config* variation from Tune's Variant Generator or a user-specified **search algorithm**. The trials are scheduled and managed by a **trial scheduler**.
+.. image:: images/tune-api.svg
 
-More information about Tune's `search algorithms can be found here <tune-searchalg.html>`__. More information about Tune's `trial schedulers can be found here <tune-schedulers.html>`__.
+More information about Tune's `search algorithms can be found here <tune-searchalg.html>`__. More information about Tune's `trial schedulers can be found here <tune-schedulers.html>`__. You can checkout out our `examples page <tune-examples.html>`__ for more code examples.
```
Comment (ericl) [inline doc/source/tune-usage.rst]: checkout out

File: doc/source/tune-usage.rst [hunk 3]
```diff
@@ -91,9 +101,43 @@ This function will report status on the command line until all Trials stop:
      - train_func_4_lr=0.4,momentum=2:  RUNNING [pid=6800], 209 s, 41204 ts, 70.1 acc
      - train_func_5_lr=0.6,momentum=2:  TERMINATED [pid=6809], 10 s, 2164 ts, 100 acc
 
-
 All results reported by the trainable will be logged locally to a unique directory per experiment, e.g. ``~/ray_results/example-experiment`` in the above example. On a cluster, incremental results will be synced to local disk on the head node.
 
+Trial Parallelism
+~~~~~~~~~~~~~~~~~
+
+Tune automatically sets the trial concurrency to N, where N is the number of CPUs (cores) on your machine. You can override the detected number of cores with ``resources_per_trial``:
```
Comment (ericl) [inline doc/source/tune-usage.rst]: I can't parse this sentences.

File: doc/source/tune-usage.rst [hunk 4]
```diff
@@ -91,9 +101,43 @@ This function will report status on the command line until all Trials stop:
...
+
+.. code-block:: python
+
+    # If you have 4 CPUs on your machine, this will run 2 concurrent trials at a time.
+    tune.run(trainable, num_samples=10, resources_per_trial={"cpu": 2})
+
+    # If you have 4 CPUs on your machine, this will run 1 trial at a time.
+    tune.run(trainable, num_samples=10, resources_per_trial={"cpu": 4})
+
+Tune can also automatically detect GPUs on your machine. To leverage GPUs, you can set ``gpu`` in ``resources_per_trial``. See more details about `using GPUs <tune-usage#using-gpus-resource-allocation>`_. Trials will only be executed if all resources are satisifed:
```
Comment (ericl) [inline doc/source/tune-usage.rst]: satisfied?

File: python/ray/tune/trainable.py (hunk 1 above)
Comment (ericl) [reply python/ray/tune/trainable.py]: Why not auto-detect these issues? (check path.startswith("/"))

Also I have no idea what validate_save_restore does, probably should remove that.

File: python/ray/tune/trainable.py [hunk 5]
```diff
@@ -489,9 +488,15 @@ def _save(self, checkpoint_dir):
     def _restore(self, checkpoint):
         """Subclasses should override this to implement restore().
 
+         The directory structure is expected to be the same as that left by
```
Comment (ericl) [inline python/ray/tune/trainable.py]: Can you provide some examples for save/restore? I find it hard to understand what the semantics are.

File: python/ray/tune/trainable.py [hunk 6]
```diff
@@ -514,7 +519,11 @@ def _log_result(self, result):
         self._result_logger.on_result(result)
 
     def _stop(self):
-        """Subclasses should override this for any cleanup on stop."""
+        """Subclasses should override this for any cleanup on stop.
+
+        If any Ray actors are launched in the Trainable (i.e., with a RLlib
+        trainer), be sure to kill the Ray actors here.
```
Comment (ericl) [inline python/ray/tune/trainable.py]: How do I kill an actor?

File: python/ray/tune/trainable.py [hunk 7]
```diff
@@ -40,14 +40,11 @@ class Trainable(object):
     Calling ``save()`` should save the training state of a trainable to disk,
     and ``restore(path)`` should restore a trainable to the given state.
 
-    Generally you only need to implement ``_train``, ``_save``, and
-    ``_restore`` here when subclassing Trainable.
+    Generally you only need to implement ``_setup``, ``_train``,
+    ``_save``, and ``_restore`` when subclassing Trainable.
```
Comment (ericl) [inline python/ray/tune/trainable.py]: reset_config?

File: python/ray/tune/trainable.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/tune/trainable.py]: Users do `pickle.save({some_info: "absolute_path"})` as an implementation detail that does not go through Tune. 

Users have already stated benefits of `validate_save_restore`, so I'm not going to remove that. https://groups.google.com/d/msg/ray-dev/pOl26QImkv4/9GhKIzOQBQAJ

File: python/ray/tune/trainable.py (hunk 5 above)
Comment (richardliaw) [reply python/ray/tune/trainable.py]: OK that's a good point..

File: python/ray/tune/trainable.py (hunk 1 above)
Comment (ericl) [reply python/ray/tune/trainable.py]: Shouldn't this say how to verify it then, with an example?

Comment (richardliaw) [reply python/ray/tune/trainable.py]: Oh I see, I thought you were pointing at the rst where there was an example; ok I’ll add that - thanks.

File: python/ray/tune/trainable.py (hunk 5 above)
Comment (richardliaw) [reply python/ray/tune/trainable.py]: still todo.

**Comment by richardliaw** [comment]
@ericl take another pass?

File: doc/source/tune-usage.rst [hunk 8]
```diff
@@ -91,9 +101,46 @@ This function will report status on the command line until all Trials stop:
      - train_func_4_lr=0.4,momentum=2:  RUNNING [pid=6800], 209 s, 41204 ts, 70.1 acc
      - train_func_5_lr=0.6,momentum=2:  TERMINATED [pid=6809], 10 s, 2164 ts, 100 acc
 
-
 All results reported by the trainable will be logged locally to a unique directory per experiment, e.g. ``~/ray_results/example-experiment`` in the above example. On a cluster, incremental results will be synced to local disk on the head node.
 
+Trial Parallelism
+~~~~~~~~~~~~~~~~~
+
+Tune automatically N concurrent trials, where N is the number of CPUs (cores) on your machine. By default, Tune assumes that each trial will only require 1 CPU. You can override this ``resources_per_trial``:
```
Comment (richardliaw) [inline doc/source/tune-usage.rst]:
```suggestion
Tune automatically N concurrent trials, where N is the number of CPUs (cores) on your machine. By default, Tune assumes that each trial will only require 1 CPU. You can override this with ``resources_per_trial``:
```

File: doc/source/tune-usage.rst [hunk 9]
```diff
@@ -91,9 +101,46 @@ This function will report status on the command line until all Trials stop:
...
+    # If you have 4 CPUs on your machine, this will run 4 concurrent trials at a time.
+    tune.run(trainable, num_samples=10)
+
+    # If you have 4 CPUs on your machine, this will run 2 concurrent trials at a time.
+    tune.run(trainable, num_samples=10, resources_per_trial={"cpu": 2})
+
+    # If you have 4 CPUs on your machine, this will run 1 trial at a time.
+    tune.run(trainable, num_samples=10, resources_per_trial={"cpu": 4})
+
+To leverage GPUs, you can set ``gpu`` in ``resources_per_trial``.  A trial will only be executed if there are resources available. See more details about `resource allocation <tune-usage#resource-allocation-using-gpus>`_, which provides more details about GPU usage and trials that are distributed:
```
Comment (richardliaw) [inline doc/source/tune-usage.rst]:
```suggestion
To leverage GPUs, you can set ``gpu`` in ``resources_per_trial``.  A trial will only be executed if there are resources available. See the section on`resource allocation <tune-usage#resource-allocation-using-gpus>`_, which provides more details about GPU usage and trials that are distributed:
```

File: doc/source/tune-usage.rst [hunk 10]
```diff
@@ -149,6 +196,22 @@ You can use ``tune.grid_search`` to specify an axis of a grid search. By default
...
+
+    tune.run(
+        ...,
+        config={
+            "alpha": tune.sample_from(lambda spec: np.random.uniform(100)),
+            "beta": tune.sample_from(lambda spec: spec.config.alpha * np.random.normal())
+        }
+    )
+
+Tune provides a couple utilities for specifying custom functions, wrapping popular numpy random utilities such as ``np.random.uniform``, ``np.random.choice``, and ``np.random.randn``. See the `Package Reference <tune-package-ref.html#ray.tune.uniform>`_ for more details.
```
Comment (richardliaw) [inline doc/source/tune-usage.rst]:
```suggestion
Tune provides a couple helper functions for common parameter distributions, wrapping numpy random utilities such as ``np.random.uniform``, ``np.random.choice``, and ``np.random.randn``. See the `Package Reference <tune-package-ref.html#ray.tune.uniform>`_ for more details.
```

File: doc/source/tune-usage.rst [hunk 11]
```diff
@@ -260,34 +326,30 @@ When running a hyperparameter search, Tune can automatically and periodically sa
...
+            return tmp_checkpoint_dir
 
-        def _restore(self, checkpoint_prefix):
-            self.saver.restore(self.sess, checkpoint_prefix)
+        def _restore(self, tmp_checkpoint_dir):
+            checkpoint_path = os.path.join(tmp_checkpoint_dir, "model.pth")
+            self.model.load_state_dict(torch.load(checkpoint_path))
 
-Checkpoints will be saved by training iteration to ``local_dir/exp_name/trial_name/checkpoint_<iter>``. You can restore a single trial checkpoint by using ``tune.run(restore=<checkpoint_dir>)``. To test if your Trainable will checkpoint and restore correctly, you can use ``tune.util.validate_save_restore`` as follows:
+Checkpoints will be saved by training iteration to ``local_dir/exp_name/trial_name/checkpoint_<iter>``. You can restore a single trial checkpoint by using ``tune.run(restore=<checkpoint_dir>)``.
```
Comment (richardliaw) [inline doc/source/tune-usage.rst]:
```suggestion
Checkpoints will be saved by training iteration to ``local_dir/exp_name/trial_name/checkpoint_<iter>``. You can restore a single trial checkpoint by using ``tune.run(restore=<checkpoint_dir>)``.

Tune also generates temporary checkpoints for pausing and switching between trials. For this purpose, it is important not to depend on absolute paths in the implementation of ``save``. See the below reference:
```

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Nice

**Comment by richardliaw** [comment]
God bless you

**Pull request merged**